### PR TITLE
Remove Java2D from Nifty Gui.

### PR DESCRIPTION
**Pull request by ghost**
The Java2D renderer is some of the most outdated, nasty code in the
whole library. I don't think anyone uses it any more.

### CODE REVIEW

**Comment by relu91** [comment]
Actually nifty-editor does .. it was the only backend that worked well with java2d and had the flexibility to shut off the input processing. I understand that is too old, is there any change to keep it updated with the upcoming changes in rendering process?

**Comment by ghost** [comment]
@relu91 I don't understand what you're saying about why you need this. Can you please be more specific? I'm sure that whatever you're doing can be accomplished in another rendering backend. And you're probably the only person brave enough to use this ancient backend. ;-)

**Comment by relu91** [comment]
Back in the days when I was building nifty-editor and choosing a renderer back end that worked with swing application. I found that the only one which works with standard swing input was Java2d. All the others that I tried eat keyboard focus when a user select on the panel where nifty-gui was rendererd. So I was basically forced to use Java2d .... I tried also all the ignore methods but none of them worked. I was far less expert so if I put my hands again on the redererer maybe I could deal with that issue, but I think it will lead to pain XD . Do you know about a success integration with swing and JME renderer ? or LWJGL ?

**Comment by ghost** [comment]
@relu91 Why do you have to use swing? The whole point of Nifty Gui is to not need it or use it. Why not use something else?

**Comment by relu91** [comment]
Are you talking about building the entire nifty-editor using only nifty-gui ? In this case it would be cool but it's even harder ... the main problem is that every gui in nifty-editor has its own instance of Nifty object and also swing has tables on nice toolbars menus. At the time was easier to use swing than built custom controls with nifty.

**Comment by ghost** [comment]
@relu91 How about having the java2d renderer in your own fork of nifty gui? I don't think anyone else uses it and it's just begging to be removed. It's very old and crusty. I'm itching to get rid of it. 

You can take a look at these:

http://hub.jmonkeyengine.org/wiki/doku.php/jme3:advanced:swing_canvas

http://stackoverflow.com/questions/12445682/rendering-javax-swing-jbutton-in-a-lwjgl-window/12447916#12447916

That would be very nifty to build Nifty with Nifty. You could try the LibGDX rendering backend. I just added a ton of flexibility for input recently: See https://github.com/void256/nifty-gui/commit/c3e2591ca92080bff98ba993069b735ede889091.

**Comment by relu91** [comment]
Yep it's ok to have java2d renderer in my forked repo! I surely won't to slow down the development due to an old renderer! 
the problem with JME and LWJGL it's that they use canvas for rendering that it's an highweight component and it doesn't mix well with other swing components... Anyway I'll try to do my best to change the rendering process in nifty-editor. 
Yeah I saw your efforts with libgdx and they are great and actually Spine was built using libgdx, I could certainlly look also in that direction but first I would add all the editing feature to the editor ... otherwise it's pretty unuseful ! Anyway thanks for the suggestions :smiley:

**Comment by ghost** [comment]
@relu91 Hey, I've got an even better idea. Why don't you put the renderer in the nifty editor code? You know it's just an implementation of Nifty's RenderDevice interface. So since you've got access to Nifty already, it's as if you wrote your own custom renderer for the Nifty Editor. You could even rename it to whatever you want, and then you would have free reign to customize it, maintain it, etc.

I'm glad you're on board to wanting to help move Nifty aggressively forward in the rendering department - it needs it badly ;-).

**Comment by relu91** [comment]
Yeah it would be even better :+1: . Sorry for not merging your pull requests but I'm always waiting for @void256 feedback since he gave me the collaborator access only to upload the wiki images.

**Comment by ghost** [comment]
@relu91 No problem, I didn't even know you could do that lol. I hope @void256 is okay. Haven't seen him for quite a while.

**Comment by oberien** [comment]
That's also my problem... You remember the background-issue and pull-request? :) That is the only main thing for us to release the pre-alphpa version of our game... :/

Hopefully he'll come back soon to merge these pull requests into the repository. As I'm not that familiar with git (only using EGit), the link you provided to include pull requests oneself doesn't really help me.

I think I'll wait until @void256 comes back and merges.

**Comment by void256** [comment]
Yeah, actually I was on vacation which was needed badly :wink: and very nice actually :smile: Sorry about that :wink: I wasn't able to do any Nifty work but I'm back now! :sunglasses: 

I'm not sure about removing the Java2D renderer yet. I agree with @3xp0n3nt that's ancient and I think @relu91 really might be the only one still using it. I haven't looked into it in years and I didn't originally wrote it myself so I'm not feeling very comfortable with it.

BUT :smile:

I always liked the idea to have a Java2D Nifty renderer. For special cases like the Nifty-Editor that @relu91 is building - and yes, I think it's the best way to integrate Nifty into Swing applications - and on the other hand to have somekind of "slow" fallback on devices or in situations that don't have support for all the other backends we have (with libGDX I doubt there is any plattform missing though).

In short: I'd really prefer if we'd be able to keep it but yes, it needs a maintainer in that case that would be able to polish / finish it.

I wouldn't be too sad if we would drop it though.

Any more opionions about this one?

**Comment by relu91** [comment]
My fear is that Java2d , since is quite old, could slow down the development of new rendering feature of nifty.. I mean if somehow we could have always a really simply backend renderer that could also be implemented with java2d without influence some other uber-render it's ok, but don't know opengl and how it will be the new rendering system since @3xp0n3nt is working on it. So I would say , yes keep it but only if  it doesn't annoying :smile: even if it will be harder to have nifty-editor working with a no-java2d renderer and latest nifty-gui lib. 

--Of topic--
JMESoc is finished and nifty-editor finally has some support for custom controls and styles this means that you can import user control in the visual editor and also apply costum styles within . I've made a little change in nifty to support that , pull request as soon as I've cleaned the code. 

p.s. Welcome back ! :smile:

**Comment by ghost** [comment]
Well, the problem is that I refuse to maintain the Java2D code. It's nasty, and I don't have any desire to mess with Java2D. But I definitely see that it could cause problems when I aggressively refactor the rendering system again. Maybe not today, maybe not tomorrow... but eventually. Not to mention there is no examples for it either, and no one has complained about that because I think anyone else is using it (there is never any comments / issues / pull requests about it anyway). It seems like a good idea to just treat it as a custom RenderDevice implementation integrated into the nifty-editor codebase.

That way, when I refactor the rendering system, if it breaks the Java2D code, then I won't be stuck in this awkward place of having to wait for @relu91 to solve the bugs before submitting rendering-related pull requests to the main Nifty codebase. If my pull request breaks @relu91's nifty-editor code, the onus would be on him to upgrade / maintain his own custom RenderDevice implementation.

@void256 What you mentioned about having a fallback renderer, I think that might have been a valid reason some years ago. But now that we have so many rendering backends, with possibly more to come (android, ios, lwjgl 3), and because Java2D is kind of the redheaded stepchild in the OpenGL neighborhood now (@void256 do you understand this Americanism? it's like the weird, unpopular choice that no one wants any more), I don't think it's really necessary. We already have the "compatibility" OpenGL "fallback" (aka non-core profile) internal renderer that works on an ancient 2006 macbook with a nasty Intel GMA 950 integrated gpu.

I personally think that's all the fallback we need.

**Comment by ghost** [comment]
@void256 I knew there would be merge conflicts in the pom.xml (because of all the other simultaneous PR's) I was gonna warn you of that. I'll fix that asap just in case you decide to merge.

**Comment by ghost** [comment]
@void256 I fixed the merge conflicts in pom.xml.

**Comment by relu91** [comment]
I'm not afraid to maintain this renderer by the way, but I see the @3xp0n3nt point. What about an other repository?? people would easly found it ... Moreover if we add also a nifty-issue repo we could collect all those repository in an organization :smiley: Would it be ok for you , @void256 ? 

p.s.

> (there is never any comments / issues / pull requests about it anyway)

 Actually it's not true @avatarscape submits two issues about Java2d renderer , of course this doesn't change anything :+1:

**Comment by void256** [comment]
@relu91 If you volunteer to help @3xp0n3nt to maintain the Java2D renderer and maybe update it to the refactorings he's planning then ... I would vote for keeping the Java2D renderer and to close this pull request without applying it.

Counterparts? :sunglasses:

**Comment by fbnz156** [comment]
The Java2D renderer is what I would consider basic, to be honest. Java's own GUI libraries all suck, except JavaFX. Regardless, It's also used in the editor project IIRC.

The only issue is... It's not being maintained.

**Comment by bgroenks96** [comment]
I will volunteer to help maintain the Java2D renderer.  I think it's a useful fallback component to have, and it might even be useful to spend some time refactoring it.

I spent a lot of time researching ways to make the best of Java2D performance a while back.  TBH I probably know it inside and out better than I do OpenGL (although that's not saying much).

**Comment by relu91** [comment]
So can we consider the Java2d renderer saved ? :smiley:  now we have two people who can maintain it :)

**Comment by fbnz156** [comment]
The Java2D renderer really needs remaking though, it's so outdated.

**Comment by void256** [comment]
OK, let's make a little vote :wink: Do we want to keep the Java2D renderer and close this pull request?

:+1:

**Comment by relu91** [comment]
:+1:

**Comment by bgroenks96** [comment]
I vote to keep it.

:+1:

**Comment by fbnz156** [comment]
:+1: 

But it should be remade. Too much doesn't work to fix it.

**Comment by ghost** [comment]
Ok, I'm closing this, but please, someone rewrite this ASAP, maybe @bgroenks96 since you know the API so well (I don't use Java2D).

**Comment by bgroenks96** [comment]
Could you give me a general idea of what needs to be rewritten and with what standards it should comply?

**Comment by fbnz156** [comment]
In general, the entire thing could do with a rewrite. Complying with the same Java standards as the main project goes without saying, and it should implement all of the features of the renderdevices, not leave some as just // TODO or whatever.

**Comment by relu91** [comment]
I've opened an other issue about this.. please continue the discussion there! :smiley:
